### PR TITLE
LMDB corruption recovery -- follow-up

### PR DESCRIPTION
**Pull request by vpodzime**
Follows #3873

Merge together:
https://github.com/cfengine/core/pull/3880
https://github.com/cfengine/enterprise/pull/556
https://github.com/cfengine/libntech/pull/27

### CODE REVIEW

File: cf-execd/cf-execd.c [hunk 1]
```diff
@@ -135,9 +135,23 @@ static const char *const HINTS[] =
...
+    bool force_repair = false;
     {
-        repair_lmdb_default();
+        char repair_flag_file[PATH_MAX] = { 0 };
+        xsnprintf(repair_flag_file, PATH_MAX, "%s%c%s",
+                  GetStateDir(), FILE_SEPARATOR, CF_DB_REPAIR_TRIGGER);
+        /* This is full of race-conditions, but it's just a best-effort
+         * thing. If a force-repair is missed, it will happen next time. If it's
+         * done twice, no big deal. */
+        if (access(repair_flag_file, F_OK) == 0)
```
Comment (olehermanse) [inline cf-execd/cf-execd.c]: Copy-pasted code, could maybe be a function?

File: cf-check/repair.c [hunk 2]
```diff
@@ -78,9 +79,52 @@ int remove_files(Seq *files)
...
+     * locally. */
+    int local_fd_tstamp = -1;
+    if (fd_tstamp == -1)
+    {
+        char *tstamp_file = StringFormat("%s.repaired", file);
+        fd_tstamp = safe_open(tstamp_file, O_CREAT|O_RDWR);
+        free(tstamp_file);
+        local_fd_tstamp = fd_tstamp;
+    }
+    if (!ExclusiveLockFileCheck(fd_tstamp) && ExclusiveLockFile(fd_tstamp, true) == -1)
```
Comment (olehermanse) [inline cf-check/repair.c]: I think the comparison expression should be parenthesized.

File: cf-agent/cf-agent.c [hunk 3]
```diff
@@ -244,9 +244,23 @@ int main(int argc, char *argv[])
     struct timespec start = BeginMeasure();
 
     GenericAgentConfig *config = CheckOpts(argc, argv);
-    if (PERFORM_DB_CHECK)
+    bool force_repair = false;
+    {
+        char repair_flag_file[PATH_MAX] = { 0 };
+        xsnprintf(repair_flag_file, PATH_MAX, "%s%c%s",
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply cf-agent/cf-agent.c]: Yup.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4006/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4006/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4006/

**Comment by olehermanse** [comment]
@cf-bottom One more jenkins, with exotics.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4017/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4017/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4017/

**Pull request merged**